apps/passport-server/src/apis/lemonade/lemonadeAPI.ts

```diff
@@ -1,78 +1,241 @@
+  LemonadeTicketSchema,
+  LemonadeTicketTypes
+} from "./types";
 
 /**
- * Used in tests.
+ * Matches the structure of Lemonade's GraphQL API.
  */
-export interface LemonadeUser {
-  id: string;
-  email: string;
-  apiKey: string;
-  name: string;
-}
+export interface ILemonadeAPI {
+  getHostingEvents(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials
+  ): Promise<LemonadeEvent[]>;
 
-/**
- * A lemonade ticket as represented by the upcoming Lemonade API. It's
- * still t.b.d so this is my best guess for now.
- */
-export interface LemonadeTicket {
-  id: string;
-  name: string;
-  email: string;
-  eventId: string;
-  tierId: string;
-  checkedIn: boolean;
-  checkerEmail?: string;
-}
+  getEventTicketTypes(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string
+  ): Promise<LemonadeTicketTypes>;
 
-/**
- * A lemonade ticket tier as represented by the upcoming Lemonade API. It's
- * still t.b.d so this is my best guess for now.
- */
-export interface LemonadeTicketTier {
-  name: string;
-  id: string;
-}
+  getTickets(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string
+  ): Promise<LemonadeTicket[]>;
 
-/**
- * A lemonade event as represented by the upcoming Lemonade API. It's
- * still t.b.d so this is my best guess for now.
- */
-export interface LemonadeEvent {
-  id: string;
-  name: string;
-  tickets: LemonadeTicket[];
-  tiers: LemonadeTicketTier[];
-  permissionedUserIds: string[];
+  checkinUser(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string,
+    lemonadeUserId: string
+  ): Promise<LemonadeCheckin>;
 }
 
 /**
- * TODO: implement to match the actual API
+ * Each Lemonade API request is made to a back-end URL, which varies only by
+ * environment (unlike Pretix, the URL is the same for all users in the same
+ * environment). The credentials are OAuth tokens, which need to be renewed
+ * periodically.
+ *
+ * Following the convention of PretixAPI, a single instance of LemonadeAPI is
+ * shared between all pipelines. This means that account-specific credentials
+ * must be passed in on each API call.
  */
-export interface ILemonadeAPI {
-  loadEvents(apiKey: string): Promise<LemonadeEvent[]>;
-  checkinTicket(
-    apiKey: string,
-    eventId: string,
-    ticketId: string
-  ): Promise<void>;
-  // TODO: fill in the other methods. This is what is planned so far:
-  // - API Key scoped to an ‘account’, which has read/write permissions to
-  //   the appropriate events. E.g. can read/write events that are owned/co-owned by the account.
-  // - GET product types for a given event
-  // - GET tickets for a given event, which will include
-  //     - Attendee name
-  //     - Attendee email
-  //     - Product type (7-day, early bird, GA, etc)
-  //     - Checked-in status
-  // - POST Check-in (and potentially check-out)
+class LemonadeAPI implements ILemonadeAPI {
```

Comment:
I've made a change to re-create the `LemonadeClient` object for each use, rather than re-using it between calls. We don't need any of the GraphQL caching features that a long-lived client would provide anyway.

Unlike the Pretix API, there is no request queue or rate-limiting, so the LemonadeAPI client is much closer to being a set of utility functions. The only shared state is the authentication token manager, which ensures that it cannot make concurrent requests for new tokens for the same credentials.